### PR TITLE
QLSpeciifcation: Add documentation for expression pragmas 

### PR DESCRIPTION
**Pull request by alexet**
The main part of this PR is adding documentation for expression pragmas. The semantics are simple but I gave a vague explanation of what the actual cases do.

I also drive-by fixed a couple of minor issues in the grammar:

- Cases missing from the local fragment for primary (but not from the dull description). Also the order differed for the cases that did exist.
- `in` not allowing set literals.

### CODE REVIEW

File: docs/codeql/ql-language-reference/ql-language-specification.rst [hunk 1]
```diff
@@ -1303,6 +1307,26 @@ The values of an ``any`` expression are those values of the expression for which
...
+Expression pragmas can be used to guide optimisation.
+
+::
+   expression_pragma ::= "pragma" "[" expression_pragma_type "]" "(" expr ")"
+
+   expression_pragma_type ::= "only_bind_out" | "only_bind_into"
+
+The values of an expression pragma are the values of the expression.
+
+The type `only_bind_out` hints that uses the result of the expression pragma should not be used to guide the evaluation of the result of the contained expression.
```
Comment (intrigus-lgtm) [inline docs/codeql/ql-language-reference/ql-language-specification.rst:1322]: > The type `only_bind_out` hints that uses the result

There is a word missing, I think.

File: docs/codeql/ql-language-reference/ql-language-specification.rst [hunk 2]
```diff
@@ -1303,6 +1307,26 @@ The values of an ``any`` expression are those values of the expression for which
...
+When checking to see that all values are bound the compiler does not assume that if the result of the expression pragma is bound then the result of the contained 
+expression is bound.
+
+The type `only_bind_into` hints that uses the result of expression pragma should not be used to guide the evaluation of the result of the expression pragma.
```
Comment (intrigus-lgtm) [inline docs/codeql/ql-language-reference/ql-language-specification.rst:1326]: Same.

File: docs/codeql/ql-language-reference/ql-language-specification.rst (hunk 1 above)
Comment (jbj) [reply docs/codeql/ql-language-reference/ql-language-specification.rst:1322]:
```suggestion
The type `only_bind_out` hints that uses of the result of the expression pragma should not be used to guide the evaluation of the result of the contained expression.
```

File: docs/codeql/ql-language-reference/ql-language-specification.rst (hunk 2 above)
Comment (jbj) [reply docs/codeql/ql-language-reference/ql-language-specification.rst:1326]:
```suggestion
The type `only_bind_into` hints that uses of the contained expression should not be used to guide the evaluation of the result of the expression pragma.
```

I think there was some copy-paste confusion here. I've suggested to make this sentence the converse of the corresponding sentence for `only_bind_out`.

File: docs/codeql/ql-language-reference/ql-language-specification.rst [hunk 3]
```diff
@@ -1303,6 +1307,26 @@ The values of an ``any`` expression are those values of the expression for which
...
+~~~~~~~~~~~~~~~~~
+
+Expression pragmas can be used to guide optimisation.
+
+::
+   expression_pragma ::= "pragma" "[" expression_pragma_type "]" "(" expr ")"
+
+   expression_pragma_type ::= "only_bind_out" | "only_bind_into"
+
+The values of an expression pragma are the values of the expression.
```
Comment (jbj) [inline docs/codeql/ql-language-reference/ql-language-specification.rst:1320]:
```suggestion
The values of an expression pragma are the values of the contained expression.
```

The term _contained expression_ is also used below and should remove some ambiguity.

**Review by jbj — COMMENTED** [review]
LGTM except these details.

**Review by ginsbach — APPROVED** [review]

**Pull request merged**